### PR TITLE
Implement reload

### PR DESCRIPTION
**Pull request by stuartnelson3**
## Motivation/summary

Support config reloading in apm-server

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [ ] Documentation has been updated

## How to test these changes

Go through [this list](https://github.com/elastic/apm-server/issues/5039#issuecomment-889228676) and verify that modifications to the config file have been picked up and applied.

## Related issues

https://github.com/elastic/apm-server/issues/5039

### CODE REVIEW

File: beater/beater.go [hunk 1]
```diff
@@ -214,6 +216,33 @@ func (bt *beater) start(ctx context.Context, cancelContext context.CancelFunc, b
...
+					apmServerCommonConfig := integrationConfig.APMServer
+					if err := reloader.reload(apmServerCommonConfig, "default", nil); err != nil {
+						fmt.Println("got err while reloading:", err)
+						continue
+					}
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
```
Comment (stuartnelson3) [inline beater/beater.go:245]: for local testing

File: beater/server.go [hunk 2]
```diff
@@ -218,14 +194,72 @@ func newGRPCServer(
...
+	s.agentcfgFetchReporter = agentcfg.NewReporter(agentcfg.NewFetcher(cfg), batchProcessor, 30*time.Second)
+	ratelimitStore, err := ratelimit.NewStore(
+		cfg.AgentAuth.Anonymous.RateLimit.IPLimit,
+		cfg.AgentAuth.Anonymous.RateLimit.EventLimit,
+		3, // burst multiplier
+	)
+	if err != nil {
+		return err
+	}
+	*s.cfg = *cfg
```
Comment (axw) [inline beater/server.go:222]: This looks like it'll create data races. What's synchronising access to the config?

**Review by axw — COMMENTED** [review]
For something that can be hot-reloaded without restarting each of the components (HTTP+gRPC server, agent config fetcher, aggregation, tail-based sampling, ...), we're going to need some synchronisation to avoid data races.

What I'm imagining is something like this:

```go
// ServerParams holds parameters for running the APM Server.
type ServerParams struct {
...

        // Config returns a channel over which the APM Server config can be received.
        //
        // The initial config will be available in the channel immediately, with the possibility
        // of dynamic config updates being received later.
        Config func() <-chan *config.Config
...
}
```

Then we would need to have each of the components call this function to obtain a channel for watching config, and update themselves independently.

Easier said than done though :)

What do you think about paring all of this back, and going back to something _similar_ but a little better than what we were doing before. On config change:
 - start a whole new "server"
 - gracefully shutdown the old one

We should be able to do this with zero downtime by using SO_REUSEPORT (e.g. using https://pkg.go.dev/github.com/libp2p/go-reuseport)

File: beater/beater.go (hunk 1 above)
Comment (stuartnelson3) [reply beater/beater.go:245]: Is this something we want or think would be valuable? Happy to make the suggested improvements, just not sure if this is something we want in the project.

**Comment by stuartnelson3** [comment]
>We should be able to do this with zero downtime by using SO_REUSEPORT (e.g. using https://pkg.go.dev/github.com/libp2p/go-reuseport). It still wouldn't be ideal as this would mean resetting aggregation and sampling, but we weren't planning to tackle that initially anyway -- we could try doing something more ambitious later.

Happy to go this route, but I originally dismissed it since it only targets (to my knowledge) linux, and while it's been around for a while (since 3.9 it seems https://lwn.net/Articles/542629/), there will be customers running older machines that fail on this :) if that's fine then I say we do it.

File: beater/beater.go (hunk 1 above)
Comment (axw) [reply beater/beater.go:245]: I'd prefer not to add it right now. If you were planning to remove it before merging then just dismiss my comment

**Comment by stuartnelson3** [comment]
Ah ok, just read the code in https://pkg.go.dev/github.com/libp2p/go-reuseport, looks like it silently fails if SO_REUSEPORT isn't available

**Comment by axw** [comment]
> Happy to go this route, but I originally dismissed it since it only targets (to my knowledge) linux, and while it's been around for a while (since 3.9 it seems https://lwn.net/Articles/542629/), there will be customers running older machines that fail on this :) if that's fine then I say we do it.

Good point about Linux < 3.9, I had forgotten about that. Looking at our [support matrix](https://www.elastic.co/support/matrix), Elastic Agent doesn't support any Linux distros that use such an old kernel -- even if apm-server does. We only need reload support when running under Elastic Agent.

Windows does have `SO_REUSEADDR`, and that does roughly the same thing (and is used by go-reuseport).

**Comment by stuartnelson3** [comment]
closed in favor of #5911

**Pull request closed without merging**